Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/rolling_updater.go

```diff
@@ -368,253 +338,290 @@ func FindSourceController(r RollingUpdaterClient, namespace, name string) (*api.
+		return oldRc, increment, nil
+	}
+	// Reduce the replica count.
+	oldRc.Spec.Replicas -= increment
+	if oldRc.Spec.Replicas < 0 {
+		oldRc.Spec.Replicas = 0
 	}
-	oldRc.Spec.Replicas -= scaleAmount
-	if oldRc.Spec.Replicas < 0 || newRc.Spec.Replicas == desired {
+	// If the new is already fully scaled and ready up to the desired size, go
+	// ahead and scale old all the way down.
+	if newRc.Spec.Replicas == desired && newReady == desired {
 		oldRc.Spec.Replicas = 0
 	}
-	fmt.Fprintf(out, "Scaling %s down to %d\n", oldRc.Name, oldRc.Spec.Replicas)
-	scaledRc, err := r.scaleAndWait(oldRc, retry, wait)
+
+	fmt.Fprintf(config.Out, "Scaling %s down to %d\n", oldRc.Name, oldRc.Spec.Replicas)
+	retryWait := &RetryParams{config.Interval, config.Timeout}
+	scaledRc, err := r.scaleAndWait(oldRc, retryWait, retryWait)
+	if err != nil {
+		return nil, 0, err
+	}
+	return scaledRc, increment, nil
+}
+
+// scalerScaleAndWait returns a scaleAndWait function which scales a
+// controller using a Scaler and a real client.
+func (r *RollingUpdater) scaleAndWaitWithScaler(rc *api.ReplicationController, retry *RetryParams, wait *RetryParams) (*api.ReplicationController, error) {
+	scalerClient := NewScalerClient(r.c)
+	scaler, err := ScalerFor("ReplicationController", scalerClient)
 	if err != nil {
+		return nil, fmt.Errorf("Couldn't make scaler: %s", err)
+	}
+	if err := scaler.Scale(rc.Namespace, rc.Name, uint(rc.Spec.Replicas), &ScalePrecondition{-1, ""}, retry, wait); err != nil {
 		return nil, err
 	}
-	return scaledRc, nil
+	return r.c.ReplicationControllers(rc.Namespace).Get(rc.Name)
 }
 
-func (r *RollingUpdater) getExistingNewRc(sourceId, name string) (rc *api.ReplicationController, existing bool, err error) {
-	if rc, err = r.c.GetReplicationController(r.ns, name); err == nil {
-		existing = true
-		annotations := rc.ObjectMeta.Annotations
-		source := annotations[sourceIdAnnotation]
-		_, ok := annotations[desiredReplicasAnnotation]
-		if source != sourceId || !ok {
-			err = fmt.Errorf("Missing/unexpected annotations for controller %s, expected %s : %s", name, sourceId, annotations)
+func (r *RollingUpdater) pollForReadyPods(interval, timeout time.Duration, oldRc, newRc *api.ReplicationController) (int, int, int, error) {
+	controllers := []*api.ReplicationController{oldRc, newRc}
+	ready := map[string]int{
+		oldRc.Name: 0,
+		newRc.Name: 0,
+	}
+	err := wait.Poll(interval, timeout, func() (done bool, err error) {
+		anyReady := false
+		for _, controller := range controllers {
+			selector := labels.Set(controller.Spec.Selector).AsSelector()
+			pods, err := r.c.Pods(controller.Namespace).List(selector, fields.Everything())
+			if err != nil {
+				return false, err
+			}
+			for _, pod := range pods.Items {
+				if api.IsPodReady(&pod) {
+					ready[controller.Name]++
+					anyReady = true
+				}
+			}
+		}
+		if anyReady {
+			return true, nil
+		}
+		return false, nil
+	})
+	oldReady := ready[oldRc.Name]
+	newReady := ready[oldRc.Name]
```

Comment:
even better to `return ready[oldRc.Name], ready[newRc.Name], oldReady + newReady, err` but that's a nit